tests/System.Text.Primitives.Tests/EncodingPerfTestsUsingEncoding.cs

```diff
@@ -0,0 +1,158 @@
+            Encoding utf8 = Encoding.UTF8;
+            int utf8Length = utf8.GetByteCount(characters);
+            var utf8Buffer = new byte[utf8Length];
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
+                        utf8.GetBytes(characters, 0, characters.Length, utf8Buffer, 0);
+            }
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [MemberData(nameof(GetEncodingPerformanceTestData))]
+        public void EncodeFromUtf32toUtf8UsingEncoding(int length, int minCodePoint, int maxCodePoint, SpecialTestCases special = SpecialTestCases.None)
+        {
+            string inputString = GenerateStringData(length, minCodePoint, maxCodePoint, special);
+            char[] characters = inputString.AsSpan().ToArray();
+            Encoding utf32 = Encoding.UTF32;
+            Encoding utf8 = Encoding.UTF8;
+            int utf32Length = utf32.GetByteCount(characters);
+            var utf32Buffer = new byte[utf32Length];
+            utf32.GetBytes(characters, 0, characters.Length, utf32Buffer, 0);
+
+            int utf8Length = utf8.GetByteCount(characters);
+            var utf8Bytes = new byte[utf8Length];
+            var utf16Chars = new char[characters.Length];
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
+                    {
+                        utf32.GetChars(utf32Buffer, 0, utf32Buffer.Length, utf16Chars, 0);
+                        utf8.GetBytes(utf16Chars, 0, utf16Chars.Length, utf8Bytes, 0);
+                    }
+            }
+        }
+
+        [Benchmark(InnerIterationCount = InnerCount)]
+        [MemberData(nameof(GetEncodingPerformanceTestData))]
+        public void EncodeFromUtf8toUtf16UsingEncoding(int length, int minCodePoint, int maxCodePoint, SpecialTestCases special = SpecialTestCases.None)
+        {
+            string inputString = GenerateStringData(length, minCodePoint, maxCodePoint, special);
+            char[] characters = inputString.AsSpan().ToArray();
+            Encoding utf8 = Encoding.UTF8;
+            int utf8Length = utf8.GetByteCount(characters);
+            var utf8Buffer = new byte[utf8Length];
+            utf8.GetBytes(characters, 0, characters.Length, utf8Buffer, 0);
+
+            Encoding utf16 = Encoding.Unicode;
+            int utf16Length = utf16.GetCharCount(utf8Buffer);
+            var utf16Buffer = new char[utf16Length];
+
+            foreach (var iteration in Benchmark.Iterations)
+            {
+                using (iteration.StartMeasurement())
+                    for (int i = 0; i < Benchmark.InnerIterationCount; i++)
+                        utf16.GetChars(utf8Buffer, 0, utf8Buffer.Length, utf16Buffer, 0);
```

Comment:
This is the wrong comparison. You are passing a buffer of UTF8 bytes and UnicodeEncoding will assume that byte buffer is filled with UTF-16 chars and process them. So, this is effectively equivalent to utf-16 to utf-16 conversion with validation.

You need to call Utf8.GetBytes instead. I suspect if you do that, you will see that the performance is equivalent to the Span-based TextEncoder.